### PR TITLE
Reference Explorer

### PR DESCRIPTION
**Pull request by rossknudsen**
Hi guys,

Have a look at this and see what you think.  There are probably a few bugs to work out.  There is an issue with RD I believe where the ActiveProject changes as the dialog is opened which causes some unexpected behaviour.

Rossco

### CODE REVIEW

File: RetailCoder.VBE/Root/RubberduckModule.cs [hunk 1]
```diff
@@ -67,6 +68,7 @@ public override void Load()
             Bind<RubberduckCommandBar>().ToSelf().InSingletonScope();
             Bind<TestExplorerModel>().ToSelf().InSingletonScope();
             Bind<IOperatingSystem>().To<WindowsOperatingSystem>().InSingletonScope();
+			Bind<RegisteredLibraryModelService>().To<RegisteredLibraryModelService>().InSingletonScope();
```
Comment (Hosch250) [inline RetailCoder.VBE/Root/RubberduckModule.cs]: Would you mind fixing your indentation?

File: RetailCoder.VBE/Root/RubberduckModule.cs [hunk 2]
```diff
@@ -369,6 +371,7 @@ private IEnumerable<IMenuItem> GetRubberduckMenuItems()
                 Kernel.Get<SettingsCommandMenuItem>(),
                 Kernel.Get<InspectionResultsCommandMenuItem>(),
                 Kernel.Get<ShowSourceControlPanelCommandMenuItem>(),
+				Kernel.Get<ReferenceBrowserCommandMenuItem>(),
```
Comment (Hosch250) [inline RetailCoder.VBE/Root/RubberduckModule.cs]: Same here.

File: RetailCoder.VBE/UI/ReferenceBrowser/ReferenceBrowser.xaml [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+
+                <TextBox HorizontalAlignment="Stretch"
+                         Margin="5"/>
+
+                <Button Command="{Binding AddVbaProjectReferenceCommand}" 
+                        Grid.Column="1"
+                        Grid.Row="0"
+                        Margin="5"
+                        Padding="5">
+                    Browse...
```
Comment (Hosch250) [inline RetailCoder.VBE/UI/ReferenceBrowser/ReferenceBrowser.xaml]: Don't hard-code the button text--you have to localize it out to RubberduckUI.resx in the UI folder.  Reference the other .xaml files to see how you reference the value.

File: RetailCoder.VBE/UI/ReferenceBrowser/ReferenceBrowserViewModel.cs [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+        private void CreateViewModelForVbaProjectReference(Reference reference)
+        {
+            var model = new VbaProjectReferenceModel(reference);
+            var vm = new RegisteredLibraryViewModel(model, _vbe.ActiveVBProject);
+            _vbaProjectReferences.Add(vm);
+        }
+
+        public void Dispose()
+        {
+            (AddVbaProjectReferenceCommand as IDisposable)?.Dispose();
```
Comment (Hosch250) [inline RetailCoder.VBE/UI/ReferenceBrowser/ReferenceBrowserViewModel.cs]: Unfortunately, we can't use C# 6 features yet.

File: RetailCoder.VBE/UI/ReferenceBrowser/ReferenceBrowserViewModel.cs [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+            {
+                if (_filePicker.ShowDialog() == DialogResult.OK)
+                {
+                    _addReferenceCallback(_filePicker.FileNames[0]);
+                }
+            }
+
+            public void Dispose()
+            {
+                _filePicker?.Dispose();
```
Comment (Hosch250) [inline RetailCoder.VBE/UI/ReferenceBrowser/ReferenceBrowserViewModel.cs]: Same here.

File: RetailCoder.VBE/UI/ReferenceBrowser/RegisteredLibraryViewModel.cs [hunk 6]
```diff
@@ -0,0 +1,170 @@
...
+
+            _isActiveReference = GetIsActiveProjectReference();
+            if (IsActiveProjectReference)
+            {
+                var reference = GetActiveProjectReferenceByFilePath(FilePath);
+                CanRemoveReference = !reference.BuiltIn;
+            }
+        }
+
+        public VbaReferenceModel Model { get; }
```
Comment (Hosch250) [inline RetailCoder.VBE/UI/ReferenceBrowser/RegisteredLibraryViewModel.cs]: Sorry, no C# 6 features.

File: RetailCoder.VBE/UI/ReferenceBrowser/RegisteredLibraryViewModel.cs [hunk 7]
```diff
@@ -0,0 +1,170 @@
...
+        {
+            get { return Model.Name; }
+        }
+
+        public bool IsActiveProjectReference
+        {
+            get { return _isActiveReference; }
+            set
+            {
+                if (value == _isActiveReference)
```
Comment (Hosch250) [inline RetailCoder.VBE/UI/ReferenceBrowser/RegisteredLibraryViewModel.cs]: Where are your braces?

File: RetailCoder.VBE/UI/ReferenceBrowser/RegisteredLibraryViewModel.cs [hunk 8]
```diff
@@ -0,0 +1,170 @@
...
+        }
+
+        public short MajorVersion
+        {
+            get { return Model.MajorVersion; }
+        }
+
+        private void AddReferenceToActiveProject()
+        {
+            //_activeProject.References.AddFromGuid(Guid.ToString("B"), MajorVersion, MinorVersion);
```
Comment (Hosch250) [inline RetailCoder.VBE/UI/ReferenceBrowser/RegisteredLibraryViewModel.cs]: Why is this commented out?

File: RetailCoder.VBE/UI/ReferenceBrowser/ReferenceBrowserViewModel.cs [hunk 9]
```diff
@@ -0,0 +1,166 @@
...
+        public ReferenceBrowserViewModel(VBE vbe, RegisteredLibraryModelService service, IOpenFileDialog filePicker)
+        {
+            _vbe = vbe;
+            _service = service;
+
+            _registeredComReferences = new ObservableCollection<RegisteredLibraryViewModel>();
+            ComReferences = new CollectionViewSource {Source = _registeredComReferences}.View;
+            //ComReferences.DeferRefresh();
+            ComReferences.SortDescriptions.Add(
+                new SortDescription(nameof(RegisteredLibraryViewModel.CanRemoveReference),
```
Comment (Hosch250) [inline RetailCoder.VBE/UI/ReferenceBrowser/ReferenceBrowserViewModel.cs]: Sorry, no C# 6 features.

**Comment by Hosch250** [comment]
No C# 6 features, please.  That means no get-only properties without a body, no null-checking `?` operators (I forget the name), and no `nameof`.

**Comment by retailcoder** [comment]
I have this PR in my fork, working on the UI and turning it into a dialog - there shouldn't be any actual action done on the active project until the dialog is okayed. Not rushing though, this is a 2.1 feature.

@rossknudsen I didn't dig too deep into that COM reflection package, but could it be used to move code out from our COM collector? Or is it just to get the list of registered COM libraries and types?

**Comment by rossknudsen** [comment]
Sorry @retailcoder , I've been really ill the past couple of weeks, which is why this hasn't gone anywhere.  I had been working on the dialog but I'm not sure I pushed everything up to my fork, so we are probably out of sync now :-(. 

It may be easier for me to start from a fresh pull from your branch rather than try and merge my changes now.  Let me know which branch you've got it in.

**Comment by rossknudsen** [comment]
Oh and to answer your other question.  The code in the COM reflection package could be used to replace some of the code in the 'COM collector'.  However, the code is pretty green.

I recommend waiting until the Reference Browser is working well, having all the bugs ironed out.  I've got a sneaking suspicion that I am doing something wrong in enumerating registry keys using `yield return`.

**Pull request closed without merging**